### ISSUE TITLE
Heredocs get confused about lexical scope when used with some block constructs

### ISSUE THREAD

**Issue by codesections** [rakudo/rakudo#4539]
The string created by a `q:to` heredoc is conceptually supposed to occur on the line in which the heredoc is started.  This means that we can correctly write code like

```raku
sub f { my $a = 'foo'; qq:to/END/ }
       $a
       END
say f; # OUTPUT: «foo»
```

This works because `$a` is evaluated in `&f`'s lexical scope, despite the fact that the _token_ `$a` occurs after `f`'s closing bracket.

However, it seems that some block contructs confuse the heredoc about this point.  Here's a golfed example:
```raku
sub f { {my $a = 'foo'; qq:to/END/ } }
       $a
       END
say f;
# OUTPUT: «===SORRY!=== Error while compiling /n Variable '$a' is not declared»
```

I've also run into this issue with `if` and `when` blocks – it seems that many nested blocks can trigger the bug shown above.

**Comment by tbrowder** [comment]
I'm surprised the first example worked, much less the second.

**Comment by codesections** [comment]
> I'm surprised the first example worked, much less the second.

Yeah, it's not _super_ well-documented at the moment, but it's a consequence of the fact that, as [the docs say](https://docs.raku.org/language/quoting#Heredocs:_:to)

> The contents of the heredoc always begin on the next line, so you can (and should) finish the line.

And the accompanying example shows that in action:

```raku
my $escaped = my-escaping-function(q:to/TERMINATOR/, language => 'html'); 
Here are the contents of the heredoc.
Potentially multiple lines.
TERMINATOR
```

I'd be happy to add more details to the docs, but don't want to do so before we debug some of the corner cases (to avoid creating an attractive nuisance).

**Comment by jnthn** [comment]
Generally Raku is very strict about lexical scopes being delimited by curly braces. Glancing at the block parsing rule in the grammar, we find this:

```
    token blockoid {
        :my $*CURPAD;
        :my %*HANDLERS;
        <.finishpad>
        :my $borg := $*BORG;
        :my $has_mystery := $*MYSTERY ?? 1 !! 0;
        { $*BORG := {} }
        [
        | '{YOU_ARE_HERE}' <you_are_here>
        | :dba('block')
            '{'
            <!!{ $*VARIABLE := '' if $*VARIABLE; 1 }>
            <statementlist(1)>
            [<.cheat_heredoc> || '}']
            <?ENDSTMT>
        || <.missing_block($borg, $has_mystery)>
        ]
        { $*CURPAD := $*W.pop_lexpad() }
    }
```

The `cheat_heredoc` thing grabs attention, although it's not so far as I can tell the reason for the variable being possible to access in the first example; in fact, its role is entirely to make sure that if a heredoc follows a `constant` declaration or `BEGIN` block then we eat it before trying the parse-time evaluation.

Instead, the reason for the variable living beyond the end of the curly brace when there's a heredoc is that we call `pop_lexpad` after we've already eaten the heredoc, so it appears still in scope I suspect we can fix the inconsistency (that is, consistently forbid it) by moving that to before the line `[<.cheat_heredoc> || '}']`, rendering the variable out of scope before we parse the heredoc content.

**Comment by jnthn** [comment]
> although it's not so far as I can tell the reason for the variable being possible to access in the first example

Argh, that wasn't quite what I wanted to express. I meant that the *intent* of `cheat_heredoc` was not to cause that variable to be in scope; this is just fallout from the `BEGIN`-time issue it is trying to solve.

**Comment by tbrowder** [comment]
I have made @jnthn's suggested change and the spectests all work. 

Now I will check if @codesections' heredoc-with-block still works or not (which should be the case according to @jnthn). If it doesn't work, then I'll add one or more tests to ensure @codesections' example and other similar examples do **not** work.

**Comment by tbrowder** [comment]
I changed to draft because @JJ reminded me to check the Speculations. There I found specific information that discusses the situation @codesections found and shows how that is supposed to be coded.  More to come...

**Comment by tbrowder** [comment]
Now the PR is ready for review.

**Comment by tbrowder** [comment]
Closed with merge of PR #4567
